src/docs/components/properties.md

```diff
@@ -7,219 +7,592 @@ contributors:
+import { Component, Prop, Watch, h } from '@stencil/core';
 
-...
+@Component({
+   tag: 'todo-list-item',
+})
 export class TodoList {
-  @Prop() name: string = 'Stencil';
+  // Mark the prop as required, to make sure it is provided when we use `todo-list-item`.
+  // We want stricter guarantees around the contents of the string, so we'll use `@Watch` to perform additional validation.
+  @Prop() thingToDo!: string;
 
-  @Watch('name')
-  validateName(newValue: string, oldValue: string) {
+  @Watch('thingToDo')
+  validateName(newValue: string, _oldValue: string) {
+    // don't allow `thingToDo` to be the empty string  
     const isBlank = typeof newValue !== 'string' || newValue === '';
+    if (isBlank) { 
+        throw new Error('thingToDo is a required property and cannot be empty') 
+    };
+    // don't allow `thingToDo` to be a string with a length of 1
     const has2chars = typeof newValue === 'string' && newValue.length >= 2;
-    if (isBlank) { throw new Error('name: required') };
-    if (!has2chars) { throw new Error('name: has2chars') };
+    if (!has2chars) {
+       throw new Error('thingToDo must have a length of more than 1 character')
+    };
   }
 }
 ```
+
+## @Prop() Options
+
+The `@Prop()` decorator accepts an optional argument to specify certain options to modify how a prop on a component
+behaves. `@Prop()`'s optional argument is an object literal containing one or more of the following fields: 
+
+```tsx
+export interface PropOptions {
+  attribute?: string;
+  mutable?: boolean;
+  reflect?: boolean;
+}
+```
+
+### Attribute Name (`attribute`)
+
+Properties and component attributes are strongly connected but not necessarily the same thing. While attributes are an
+HTML concept, properties are a JavaScrip concept inherent to Object-Oriented Programming.
+
+In Stencil, the `@Prop()` decorator applied to a **property** will instruct the Stencil compiler to also listen for
+changes in a DOM attribute.
+
+Usually, the name of a property is the same as the attribute, but this is not always the case. Take the following
+component as example:
+
+```tsx
+import { Component, Prop, h } from '@stencil/core';
+// `MyHttpService` is an `Object` in this example
+import { MyHttpService } from '../some/local/directory/MyHttpService';
+
+@Component({
+   tag: 'todo-list-item',
+})
+export class ToDoListItem {
+   @Prop() isComplete: boolean;
+   @Prop() thingToDo: string;
+   @Prop() myHttpService: MyHttpService;
+}
+```
+
+This component has **3 properties**, but the compiler will create **only 2 attributes**: `is-complete` and
+`thing-to-do`.
+
+```html
+<todo-list-item is-complete="false" thing-to-do="Read Attribute Naming Section of Stencil Docs"></my-cmp>
+```
+
+Notice that the `myHttpService` type is not a primitive (e.g. not a `number`, `boolean`, or `string`). Since DOM
+attributes can only be strings, it does not make sense to have an associated DOM attribute called "my-http-service".
```

Comment:
Might be helpful to show how you can still set the value of `myHttpService`.

Reply:
I was a little intentionally vague here to avoid scope creep - I stubbed `MyHttpService` as an empty class earlier in this doc, and didn't want to get into all the details as a part of this PR. How would you feel about us doing that as a follow on effort?

Reply:
I think that's totally fair. I'm notoriously bad at avoiding scope creep so I admire your discipline.